runtime/gc_glue_java/ScavengerDelegate.cpp

```diff
@@ -288,6 +290,179 @@ MM_ScavengerDelegate::internalGarbageCollect_shouldPercolateGarbageCollect(MM_En
+				if (!referentMustBeCleared && shouldScavengeReferenceObject && !_shouldScavengePhantomReferenceObjects) {
+					_shouldScavengePhantomReferenceObjects = true;
+				}
+				break;
+			default:
+				Assert_MM_unreachable();
+			}
+
+			GC_SlotObject referentPtr(env->getOmrVM(), J9GC_J9VMJAVALANGREFERENCE_REFERENT_ADDRESS(env, objectPtr));
+			if (referentMustBeCleared) {
+				/* Discovering this object at this stage in the GC indicates that it is being resurrected. Clear its referent slot. */
+				referentPtr.writeReferenceToSlot(NULL);
+				/* record that the reference has been cleared if it's not already in the cleared or enqueued state */
+				if (!isReferenceCleared) {
+					J9GC_J9VMJAVALANGREFERENCE_STATE(env, objectPtr) = GC_ObjectModel::REF_STATE_CLEARED;
+				}
+			} else if (shouldScavengeReferenceObject) {
+				env->getGCEnvironment()->_referenceObjectBuffer->add(env, objectPtr);
+			}
+
+			fomrobject_t *referentSlotAddress = referentMustBeMarked ? NULL : referentPtr.readAddressFromSlot();
+			objectScanner = GC_ReferenceObjectScanner::newInstance(env, objectPtr, referentSlotAddress, allocSpace, flags);
+		} else {
+			objectScanner = GC_MixedObjectScanner::newInstance(env, objectPtr, allocSpace, flags);
+		}
+		break;
+	case GC_ObjectModel::SCAN_OWNABLESYNCHRONIZER_OBJECT:
+		if (GC_ObjectScanner::isHeapScan(flags)) {
+			private_addOwnableSynchronizerObjectInList(env, objectPtr);
+		}
+		objectScanner = GC_MixedObjectScanner::newInstance(env, objectPtr, allocSpace, flags);
+		break;
+	case GC_ObjectModel::SCAN_CONTINUATION_OBJECT:
+		*shouldRemember = doContinuationObject(env, objectPtr, reason);
+		objectScanner = GC_MixedObjectScanner::newInstance(env, objectPtr, allocSpace, flags);
+		break;
+	case GC_ObjectModel::SCAN_POINTER_ARRAY_OBJECT:
+		{
+			uintptr_t splitAmount = 0;
+			if (!GC_ObjectScanner::isIndexableObjectNoSplit(flags)) {
+				splitAmount = _extensions->scavenger->getArraySplitAmount(env, _extensions->indexableObjectModel.getSizeInElements((omrarrayptr_t)objectPtr));
+			}
+			objectScanner = GC_PointerArrayObjectScanner::newInstance(env, objectPtr, allocSpace, flags, splitAmount);
+		}
+		break;
+		case GC_ObjectModel::SCAN_FLATTENED_ARRAY_OBJECT:
+		{
+			Assert_MM_true(J9_IS_J9CLASS_FLATTENED(clazzPtr));
+			uintptr_t slotsToDo = 0;
+			uintptr_t startIndex = 0;
+			objectScanner = GC_FlattenedArrayObjectScanner::newInstance(env, objectPtr, allocSpace, GC_ObjectScanner::indexableObject | GC_ObjectScanner::indexableObjectNoSplit, slotsToDo, startIndex);
+		}
+		break;
+	case GC_ObjectModel::SCAN_PRIMITIVE_ARRAY_OBJECT:
+		break;
+	default:
+		Assert_GC_true_with_message(env, false, "Bad scan type for object pointer %p\n", objectPtr);
+	}
+
+	return objectScanner;
+}
+
 GC_ObjectScanner *
 MM_ScavengerDelegate::getObjectScanner(MM_EnvironmentStandard *env, omrobjectptr_t objectPtr, void *allocSpace, uintptr_t flags)
```

Comment:
I think this variant is to be eventually removed, so please provide a todo comment